src/Microsoft.TestPlatform.Common/Filtering/FastFilter.cs

```diff
@@ -0,0 +1,161 @@
+            
+            this.FilterPropertyName = filterPropertyName;
+            this.FilterPropertyValues = filterPropertyValues;
+
+            if (filterOperation == Operation.Equal && (filterOperator == Operator.Or || filterOperator == Operator.None))
+            {
+                IsFilteredOutWhenMatched = false;
+            }
+            else if (filterOperation == Operation.NotEqual && (filterOperator == Operator.And || filterOperator == Operator.None))
+            {
+                IsFilteredOutWhenMatched = true;
+            }
+            else
+            {
+                throw new ArgumentException();
+            }
+        }
+
+        internal string[] ValidForProperties(IEnumerable<string> properties)
+        {
+            return properties.Contains(this.FilterPropertyName, StringComparer.OrdinalIgnoreCase)
+                ? null
+                : new[] { this.FilterPropertyName };
+        }
+
+        internal bool Evaluate(Func<string, Object> propertyValueProvider)
+        {
+            ValidateArg.NotNull(propertyValueProvider, "propertyValueProvider");
+
+            if (!TryGetSinglePropertyValue(this.FilterPropertyName, propertyValueProvider, out var value))
+            {
+                return false;
+            }
+
+            if (PropertyValueRegex != null)
+            {
+                var match = PropertyValueRegex.Match(value);
+                if (match.Success)
+                {
+                    value = match.Value;
+                }
+                else
+                {
+                    return false;
+                }
+            }
+
+            var matched = this.FilterPropertyValues.Contains(value);
+            return IsFilteredOutWhenMatched ? !matched : matched;
+        }
+
+        private static bool TryGetSinglePropertyValue(string name, Func<string, Object> propertyValueProvider, out string singleValue)
+        {
+            singleValue = propertyValueProvider(name) as string;
+            return singleValue != null;
+        }
+
+        internal static Builder CreateBuilder()
+        {
+            return new Builder();
+        }
+
+        internal sealed class Builder
+        {
+            private bool operatorEncountered = false;
+            private Operator fastFilterOperator = Operator.None;
+
+            private bool conditionEncountered = false;
+            private Operation fastFilterOperation;
+            private string filterPropertyName;
+            private ImmutableHashSet<string>.Builder filterHashSetBuilder = ImmutableHashSet.CreateBuilder<string>();
+
+            private bool containsValidFilter = true;
+
+            internal bool ContainsValidFilter => containsValidFilter && conditionEncountered;
+
+            internal void AddOperator(Operator @operator)
+            {
+                if (!containsValidFilter || !(@operator == Operator.And || @operator == Operator.Or))
+                {
```

Comment:
should `containsValidFilter` be set to false if `!(@operator == Operator.And || @operator == Operator.Or)) ` is satisfied?

Reply:
You are right.